### ISSUE TITLE
hi ! dose resample layer have back forward? if not, how the loss transport to earlier layers?  

### ISSUE THREAD

**Issue by leonardodora** [lmb-freiburg/flownet2#150]
if not, how the loss of sd-net transports to earlier layers? 
and if have, I don't find the implentation in your source.

btw, I have another question.
the Eltwise layers have done to flow for what? I don't get the point.

**Comment by nikolausmayer** [comment]
The "SD" network is a separate stream. Its loss is not backpropagated to the CSS stack. See also Figure 2 in [the paper](https://lmb.informatik.uni-freiburg.de/Publications/2017/IMSKDB17/paper-FlowNet_2_0__CVPR.pdf).

Due to its resample layers, the network needs input images with widths and heights that are multiples of 64. We rescale the inputs so they fit this need, but this means that the output of the network is not valid for the original image. It needs to be rescaled. This is what the `Eltwise` layer on the output does.

**Comment by leonardodora** [comment]
yes,  "SD" network is a separate stream, but "sd" also need to be trained. sorry，but I still don't get it how it will be trained without backpropagated？

**Comment by nikolausmayer** [comment]
As far as I can tell, only the full "FlowNet2" model uses `Resample` layers during training, and by that point not the entire stack is trained (so it's ok if `Resample` does not do backward).

All the other models, including "FlowNet2-SD", only have `Resample` layers for deployment (i.e. for testing). The `train.prototxt` files for network training/finetuning do not contain `Resample` layers because we use cropping instead.

TL;DR: "SD" is _deployed_ with Resample, but _trained_ without.

**Comment by nikolausmayer** [comment]
(closed due to inactivity)
